ArduCopter/mode_zigzag.cpp

```diff
@@ -0,0 +1,316 @@
+}
+
+// called by ZIGZAG case in RC_Channel.cpp
+// used to record point A, B and give the signal to fly to next destination automatically
+void Copter::ModeZigZag::receive_signal_from_auxsw(RC_Channel::aux_switch_pos_t aux_switch_position)
+{
+    // define point A and B
+    if (stage == REQUIRE_A || stage == REQUIRE_B) {
+        if (aux_switch_position != RC_Channel::aux_switch_pos_t::MIDDLE) {
+            Vector3f cur_pos = inertial_nav.get_position();
+            set_destination(cur_pos, aux_switch_position);
+            return;
+        }
+    } else {      
+        // A and B have been defined
+        if (aux_switch_position != RC_Channel::aux_switch_pos_t::MIDDLE) { //switch position in HIGH or LOW
+            // calculate next point A or B
+            // need to judge if the drone's position is between A and B
+            Vector3f next_dest;
+            if(calculate_next_dest(next_dest, aux_switch_position)){
+                // initialise waypoint and spline controller
+                wp_nav->wp_and_spline_init();
+                set_destination(next_dest, aux_switch_position);
+                // initialise yaw
+                auto_yaw.set_mode_to_default(false);
+                stage = AUTO;
+                zigzag_is_between_A_and_B = false;
+            }
+        } else {      //switch in middle position, regain the control
+            if (stage == AUTO) {
+                stage = MANUAL_REGAIN;
+                loiter_nav->init_target();
+                zigzag_is_between_A_and_B = true;
+            } else {
+                zigzag_is_between_A_and_B = false;
+            }
+        }
+    }
+}
+
+// set_destination - sets zigzag mode's target destination
+// Returns true if the fence is enabled and guided waypoint is within the fence
+// else return false if the waypoint is outside the fence
+bool Copter::ModeZigZag::set_destination(const Vector3f& destination, RC_Channel::aux_switch_pos_t aux_switch_position)
+{
+
+#if AC_FENCE == ENABLED
+    // reject destination if outside the fence
+    Location_Class dest_loc(destination);
+    if (!copter.fence.check_destination_within_fence(dest_loc)) {
+        copter.Log_Write_Error(ERROR_SUBSYSTEM_NAVIGATION, ERROR_CODE_DEST_OUTSIDE_FENCE);
+        // failure is propagated to GCS with NAK
+        return false;
+    }
+#endif
+    switch (stage) {
+        case REQUIRE_A:
+            //define point A
+            zigzag_waypoint.A_pos.x = destination.x;
+            zigzag_waypoint.A_pos.y = destination.y;
+            zigzag_waypoint.switch_pos_A = aux_switch_position;
+            stage = REQUIRE_B;     //next need to define point B
+            return true;
+        case REQUIRE_B:
+            //point B will only be defined after A is defined
+            //if user toggle to the switch position that were previously defined as A
+            //exit the function and do nothing
+            if(aux_switch_position == zigzag_waypoint.switch_pos_A){    
```

Comment:
nitpick: there should be a space between "if" and "(" and also between ")" and "{"